[FIR] Resolve annotations inside contracts

isuckatcs opened the pull request:
^KT-68893 Fixed

Inside a contract annotations are not resolved which results in an internal error.

I might be missing something, but seemingly both `contract` and `bar` are functions in the language, so I'm not sure why we need a different rule for annotations within contracts.
```kotlin
fun bar(block: () -> Unit) {
    block();
}

fun foo() {
    contract {
        @foo // fun contact() -> @foo
    }
    
    bar {
        @foo // fun bar() -> @foo
    }
}
```

I haven't seen any new tests failing by removing the overrides. It might be possible though that now the CFG contrains edges, it shouldn't. In that case I guess, we can just resolve the type of the annotation in the overridden methods and return.

demiurg906 commented:
Hi @isuckatcs. Thank you for the contribution, but your change is incorrect

`FirContractResolveTransformer` is used to transform all elements in file, and it is intended to transform only contract declarations on functions, without touching any other elements in the tree. So if you remove those overloads for `transformAnnotation`, annotations outside contracts will start to be transformed on `CONTRACTS` phase instead of `BODY_RESOLVE`/`IMPLICIT_BODY_RESOLVE`

So the proper solution would be to differentiate annotations inside contract blocks and outside of them

Also it's incorrect to just resolve the annotation regularly as on body resolve phase, as on contracts phase there is no guarantee that all callables have return types resolved, so code like this most likely will fail (probably in a non-trivial way)

```kotlin
annotation class Ann(val x: String)

fun foo(b: Boolean) {
    contract {
        @Ann(s) returns() implies b
    }
}

const val s = "hello" // no explicit return type
```

Potentially it can be solved by special resolution mechanism for arguments of such annotations, which will just replace all of them with error nodes with proper diagnostics. If you want, you can try to do it

demiurg906 commented:
There is one moment about contributions to the compiler: by default [we consider](https://github.com/JetBrains/kotlin/blob/2ea0cdf46de2222ae9fd49423cf4c821d95749a5/docs/contributing.md#L30) only tickets with `up-for-grabs` tag. If you feel that there is some issue which you really can and want to fix which don't contain this tag, please discuss it in the issue itself before starting a pull request (there are might be some non-trivial constraints which are not obvious at first glance if you don't have enough context). Otherwise your PR may be rejected

demiurg906 reviewed (changes requested):
*I've described everything above, but github doesn't allow me to press "Submit review"  without an additional comment

isuckatcs commented:
Hi!

Sorry for the late reply, I'm a bit busy these weeks.

> Potentially it can be solved by special resolution mechanism for arguments of such annotations, which will just replace all of them with error nodes with proper diagnostics. If you want, you can try to do it

I will definitely take a look once I manage to allocate some time for it. If I understand it correctly, we might need the result of the `BODY_RESOLVE` phase to be able to properly reason about some annotations inside contracts. That technically means introducing one more pass to revisit the by that time still unresolved contracts, right?

demiurg906 commented:
> That technically means introducing one more pass to revisit the by that time still unresolved contracts, right?

Theoretically, it can be done this way, but please don't. Such a small problem isn't worth introducing an extra resolution phase. The easiest solution probably is to have two modes for annotation resolution in `FirAbstractContractResolveTransformerDispatcher`, which do different things depending on whether we are inside the contract or outside of it:
- outside do nothing (like now)
- inside just resolve every annotation argument to error expression

It is not the best solution from a UX point of view, but annotations inside contracts are prohibited anyway, so there is no need for meaningful diagnostics about annotation arguments inside a contract to fix the problem

isuckatcs commented on compiler/fir/raw-fir/light-tree2fir/src/org/jetbrains/kotlin/fir/lightTree/converter/LightTreeRawFirExpressionBuilder.kt:
```diff
@@ -498,7 +498,7 @@ class LightTreeRawFirExpressionBuilder(
         }
 
-        val result = firExpression ?: buildErrorExpression(null, ConeNotAnnotationContainer("???"))
+        val result = firExpression ?: buildErrorExpression(annotatedExpression.toFirSourceElement(), ConeNotAnnotationContainer("???"))
```
This must have a location, otherwise we get a crash in `FirContractChecker`.

```kotlin
object FirContractChecker {
    ...
    private fun checkUnresolvedEffects(
        ...
    ) {
        ...
            reporter.reportOn(unresolvedEffect.source, ...)
                                           // ^ source is asserted to be non-null inside reportOn()
        ...
    }
    ...
}
```

demiurg906 replied:
Yep, this is a correct fix. Could you please check that the corresponding place in `PsiRawFirBuilder` also correctly sets the source?

isuckatcs replied:
Yes, it sets it. In fact I checked that first and that was my leading clue that this one should be set too.

```kotlin
val result = rawResult as? FirAnnotationContainer
    ?: buildErrorExpression(
        expression.toFirSourceElement(),
        ConeNotAnnotationContainer(rawResult?.render() ?: "???")
    )
```

isuckatcs commented on compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/transformers/contracts/FirAbstractContractResolveTransformerDispatcher.kt:
```diff
@@ -61,12 +62,25 @@ abstract class FirAbstractContractResolveTransformerDispatcher(
+            val errorType = ConeErrorType(typeDiagnostic)
+
+            annotationCall.replaceAnnotationTypeRef(errorType.toFirResolvedTypeRef(annotationCall.annotationTypeRef.source))
+        }
```
Here I was also experimenting with returning a `FirErrorAnnotationCall` with a custom error message that says `"Annotations calls are prohibited inside contract descriptions"`, though the message ended up being a duplicate of what is reported by `FirContractChecker`.

Also, since `FirErrorAnnotationCall` still needs a resolved type to not crash, we would still see the current error, so I ended up sticking to creating the error type only without changing the node too.

Is any additional change desired here?

demiurg906 replied:
Your approach is correct

Message in `ConeSimpleDiagnostic` is mostly used for debug purposes

isuckatcs replied:
I meant an error message that we put into `FirErrorsDefaultMessages.kt`.

Right now the compiler output is the following:
```
e: Error in contract description: illegal element in contract description
e: Inference error.
e: Syntax error: Expecting an expression.
```

With my message it was
```
e: Error in contract description: illegal element in contract description
e: Annotation calls are prohibited inside contract descriptions.
e: Syntax error: Expecting an expression.
```
I felt like the first 2 messages are essentially the same, but now that I think about it once again,
they are probably not.

demiurg906 commented on compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/transformers/contracts/FirAbstractContractResolveTransformerDispatcher.kt:
```diff
@@ -61,12 +62,25 @@ abstract class FirAbstractContractResolveTransformerDispatcher(
+        if (insideContractDescription) {
+            val typeDiagnostic = ConeSimpleDiagnostic(
+                "Cannot infer annotation call type during CONTRACTS phase.",
+                DiagnosticKind.InferenceError
```
Could you please introduce a new `DiagnosticKind` here, like `DiagnosticKind.AnnotationInContract`?
`InferenceError` is used for errors in type inference, and there is no inference here

isuckatcs commented:
I introduced a new error message and also renamed `DiagnosticKind.AnnotationNotAllowed` as I think it was misleading. The renaming is in it's own commit though, so it can be reverted easily.

demiurg906 reviewed (commented):
Everything looks fine, thank you

I've sent the MR to remote run
